Enhanced Dockerfile for cross compilation

crazy-max opened the pull request:
follow-up https://github.com/moby/moby/pull/43474#discussion_r846827598
closes #43492
closes #43613
closes #41590
closes #41614 (should be already fixed atm)
fixes #41862
fixes #41519
fixes #38307 (`make binary` instead of `make all`)
fixes #39380

***- What I did**

Better support for cross compilation so we can rely on `--platform` flag of buildx for a seamless integration. This removes unnecessary extra cross logic in the Dockerfile as well as reducing hack scripts. Tried my best to reduce the footprint of changes but modifying one bit in the `Dockerfile` involves other changes in `./hack` scripts. Non-sandboxed build invocation is still supported.

dev stages have been updated accordingly to changes for cross comp as well as linked tools (swagger, tomll, gotestsum, ...)

frozen-images stage doesn't use the download-frozen-image-v2.sh anymore so we can effectively use `TARGETPLATFORM` from global scope. The test util has been updated accordingly. Will be able to fix:

fixes #42701
fixes #31320
closes #31435

In a follow-up we can remove download-frozen-image-v2.sh script but needs to look first at `Dockerfile.e2e` which seems not used anymore in our pipeline.

vpnkit stage only supports linux/amd64 and linux/arm64 platforms when building dev image and will crash if we try building against another platform. With this change we can still build the dev image against any platform using dummy scratch base.

pin criu for better reproducibility and build from source so we can use it across any platform. will also fix issues with repo mirror we often encounter: https://github.com/moby/moby/pull/43682#issuecomment-1145116748

containerd, runc, tini and rootlesskit builds in Dockerfile are limited to host platform and could not be cross-built for other platforms. This change allows to build against any platforms if we want to smoke test in a follow-up but also enhance e2e tests for linux and windows in our pipeline. This also introduced `DOCKER_LINKMODE` to be able to build dynamic or static binaries.

To add support for `riscv64` builds we need crossbuild packages for `riscv64` but current `golang` image with debian bullseye does not support it. Ubuntu 22.04 supports riscv64 but unfortunately drops support for armel arch. Therefore we need a multi base image that will be picked up based on the target platform we want to build.

The current bake definition has been updated to take the changes into account as well as the ci gha workflow.

cc @tonistiigi @thaJeztah @cpuguy83 @tianon 

**- How to verify it**
```shell
# build binaries for the current host platform
# output to ./bundles/binary by default
docker buildx bake

# build binaries for the current host platform
# output to ./bin
DESTDIR=./bin docker buildx bake

# build dynamically linked binaries
# output to ./bundles/dynbinary by default
DOCKER_LINKMODE=dynamic docker buildx bake

# build binaries for all supported platforms
docker buildx bake binary-cross

# build binaries for a specific platform
docker buildx bake --set *.platform=linux/arm64

# build all for the current host platform (binaries + containerd, runc, tini, ...)
# output to ./bundles/all by default
docker buildx bake all

# build all for the current host platform
# output to ./bin
DESTDIR=./bin docker buildx bake all

# build all for all supported platforms
docker buildx bake all-cross
```

**- Description for the changelog**

* Enhanced Dockerfile for cross compilation

crazy-max commented on Dockerfile.cross:
```diff
@@ -0,0 +1,297 @@
+
+FROM containerd-base-${GO_LINKMODE} AS containerd-base-alpine
+ARG TARGETPLATFORM
+RUN xx-apk add --no-cache musl-dev gcc g++ pkgconf btrfs-progs-dev libsecret-dev && xx-go --wrap
```
we could add a mount cache here for alpine pkgs s well as for the debian variant

crazy-max commented on Dockerfile.cross:
```diff
@@ -0,0 +1,297 @@
+    libudev-dev
+
+FROM build-${BASE_VARIANT} AS build
+COPY --from=gowinres /out/ /usr/local/bin
+COPY --from=tini /out/ /out
+COPY --from=runc /out/ /out
+COPY --from=containerd /out/ /out
+COPY --from=rootlesskit /out/ /out
+COPY --from=containerutility /out /out
+COPY --from=vpnkit / /out
```
We could link copies with `COPY --link` to reuse already built layers in subsequent builds.

tonistiigi commented on Dockerfile.cross:
```diff
@@ -0,0 +1,297 @@
+FROM vpnkit-${TARGETOS} AS vpnkit
+
+FROM build-base-alpine AS build-alpine
+RUN apk add --no-cache lld
```
why are the lld's installed separately from other things like `clang` etc

tonistiigi commented on Dockerfile.cross:
```diff
@@ -0,0 +1,297 @@
+
+FROM build-${BASE_VARIANT} AS build
+COPY --from=gowinres /out/ /usr/local/bin
+COPY --from=tini /out/ /out
```
Why are these copied to the build stage. Should only by to binary stage.

crazy-max replied:
Because we want to generate the checksums file for them. Checksums fiel is created by `./hack/build/release`. Maybe we could find a better way and have a stage that does that job instead.

tonistiigi replied:
Should not be in the same step that runs `go build`

crazy-max replied:
Moved in a dedicated stage

tonistiigi commented on Dockerfile.cross:
```diff
@@ -0,0 +1,297 @@
+    --mount=type=tmpfs,target=cli/winresources/dockerd \
+    --mount=type=tmpfs,target=cli/winresources/docker-proxy \
+    --mount=type=cache,target=/root/.cache \
+    xx-go --wrap && OUTPUT=/out ./hack/build/release
```
`xx-verify`

crazy-max replied:
It's done in release script: https://github.com/moby/moby/pull/43529/files#diff-5a394e70534b6f2ab7d8356d2d88802cba6c2a298de8273444244c68723ac08bR12-R22

Wanted to add it there but there is a little too much logic to use here docs in the dockerfile I think.

tonistiigi replied:
I think people want to keep `xx` out of the scripts to the script can be used on their own on a manually prepared system.

crazy-max replied:
> I think people want to keep `xx` out of the scripts

I agree with you, that's why I check if `xx-verify` is available in the script so non-sandboxed invocation is still possible. But yeah `xx` purpose is to work inside a container. Will see how I can improve that.

crazy-max replied:
Moved `xx-verify` in the build stage.

tonistiigi reviewed (commented):
Is the support for multiple distros and static/dynamic a requirement?

tonistiigi commented on Dockerfile.cross:
```diff
@@ -0,0 +1,297 @@
+    && git checkout -q "$CONTAINERUTILITY_VERSION" \
+    && CC="$(xx-info)-gcc" CXX="$(xx-info)-g++" make \
+    && mkdir /out \
+    && mv containerutility.exe /out/
```
Is this binary something that ships with `dockerd`?

crazy-max replied:
Oh was thinking it was bundled in https://download.docker.com/win/static/stable/x86_64/ but doesn't look like it. @thaJeztah what's the purpose of it?

tonistiigi commented on Dockerfile.cross:
```diff
@@ -0,0 +1,297 @@
+    --mount=type=cache,target=/root/.cache \
+    git fetch origin \
+    && git checkout -q "$TINI_VERSION" \
+    && cmake -DCMAKE_C_COMPILER=$(xx-info)-gcc . \
```
Use `clang` and `cmake-defines` from xx

crazy-max replied:
I tried but doesn't work with clang :persevere:. I will check again.

crazy-max replied:
Ok works fine now.

tonistiigi commented on Dockerfile.cross:
```diff
@@ -0,0 +1,297 @@
+
+FROM tini-base AS tini-base-alpine
+RUN apk add --no-cache lld
+ENV XX_CC_PREFER_LINKER=ld
```
If this is needed then don't need to install `lld`

crazy-max commented:
> Is the support for multiple distros and static/dynamic a requirement?

Not necessary imo so if debian variant is the one we want to rely on, I can remove the BUILD_VARIANT logic.

cpuguy83 commented:
What's the goal of this? Who is going to use it?
The only purpose the current cross target serves, to my knowledge, is to make sure we don't break compilation on other platforms... maybe it's used for static builds?

I'm all for refactoring/rewriting our build system, I don't think anyone really likes it.
That said, the caveat I would say here is that the build system should have _no_ dependency on docker or its components except as a means of setting up a build environment.
I also do not think it should depend on obscure tools like `xx`.

Let CC and CXX be, and if someone wants to use `xx` they could do something like a hypothetical `CC=$(xx-env CC)` to get the correct compiler.
Dockerfile can use `xx`, scripts should use only standard tooling.

All in all, this looks like packaging. As a packager, I cannot use this toolchain.

crazy-max commented:
> What's the goal of this? Who is going to use it? The only purpose the current cross target serves, to my knowledge, is to make sure we don't break compilation on other platforms... maybe it's used for static builds?

The primary goal is to improve the maintainability of our build scripts. The cross compilation is just a first step, dissociated for the moment from what is used by Jenkins or the Makefile but later we should not depend on the multitude of scripts present in the hack folder (or at least a large amount of them).

> I'm all for refactoring/rewriting our build system, I don't think anyone really likes it. That said, the caveat I would say here is that the build system should have _no_ dependency on docker or its components except as a means of setting up a build environment. I also do not think it should depend on obscure tools like `xx`.
>
> Let CC and CXX be, and if someone wants to use `xx` they could do something like a hypothetical `CC=$(xx-env CC)` to get the correct compiler. Dockerfile can use `xx`, scripts should use only standard tooling.

It doesn't need docker, and `xx` is only used inside the Dockerfile. You can invoke `./hack/build/release` to build `dockerd` and `docker-proxy` with your local env. See https://github.com/moby/moby/pull/43529/files#diff-de4586e60f6c74e2d92477c3c7a7cfc1d3a8527260e6e179b2275f1bf1e6f2f3R21-R28

crazy-max commented:
> All in all, this looks like packaging.

I agree that the `binaries` stage in `Dockerfile.cross` currently outputs project binaries and also external tools. I have changed that so now it will just build binaries linked to the project (dockerd and docker-proxy). Another stage named `bundle` will output binaries and linked tools. Is it what you mean by packaging?

tonistiigi commented on hack/build/binary:
```diff
@@ -0,0 +1,135 @@
+	esac
+fi
+
+if [ -x "$(command -v xx-info)" ]; then
```
This block should be removed and shouldn't be needed for anything. Also, avoid using `gcc` in general.

crazy-max replied:
Ah indeed I missed that part. Now I see why @cpuguy83 was talking about `CC` and `CXX`.

crazy-max commented on Dockerfile.cross:
```diff
@@ -0,0 +1,292 @@
+    --mount=type=tmpfs,target=cli/winresources/dockerd \
+    --mount=type=tmpfs,target=cli/winresources/docker-proxy \
+    --mount=type=cache,target=/root/.cache \
+    # FIXME: xx doesn't seem to set CC/CXX for windows even if available
+    CC=$(xx-info)-gcc CXX=$(xx-info)-g++ OUTPUT=/out ./hack/build/release && \
```
https://github.com/tonistiigi/xx/pull/64 should fix it.

crazy-max replied:
fixes https://github.com/tonistiigi/xx/releases/tag/v1.1.1

thaJeztah commented:
Opened https://github.com/moby/moby/pull/43613 as a draft to run the Jenkinsfile changes

thaJeztah commented on Dockerfile:
```diff
@@ -278,141 +483,146 @@ RUN ldconfig
+    gcc \
+    g++ \
+    pkg-config \
+    dpkg-dev \
```
Also see other comments; can sort this list?

thaJeztah reviewed (commented):
(Just some quick comments from glancing over the Dockerfile)

crazy-max commented on Dockerfile:
```diff
@@ -1,274 +1,479 @@
+  export TINI_TARGET=$([ "$DOCKER_LINKMODE" = "static" ] && echo "tini-static" || echo "tini") \
+  && git fetch origin \
+  && git checkout -q "$TINI_VERSION" \
+  && CC=$(xx-info)-gcc cmake . \
```
Yes sounds better with heredocs :+1:

crazy-max commented on Dockerfile:
```diff
@@ -1,274 +1,479 @@
+FROM containerutility-base AS containerutility-build
+ARG CONTAINERUTILITY_VERSION
+RUN --mount=from=containerutility-src,src=/usr/src/containerutility,rw \
+  --mount=type=cache,target=/root/.cache \
```
Sure done

crazy-max commented on Dockerfile:
```diff
@@ -1,274 +1,479 @@
+ARG TARGETPLATFORM
+RUN --mount=type=cache,sharing=locked,id=moby-tini-aptlib,target=/var/lib/apt \
+    --mount=type=cache,sharing=locked,id=moby-tini-aptcache,target=/var/cache/apt \
+  xx-apt-get update && xx-apt-get install -y libc6-dev gcc
```
Done

crazy-max commented on Dockerfile:
```diff
@@ -1,274 +1,479 @@
+  --mount=type=cache,sharing=locked,id=moby-containerutility-aptcache,target=/var/cache/apt \
+  xx-apt-get install -y binutils gcc g++ dpkg-dev pkg-config
+
+FROM containerutility-base AS containerutility-build
```
not needed, was just to have a clear separation but I agree with you that some of them can be merged together. Done.

crazy-max commented on Dockerfile:
```diff
@@ -1,274 +1,479 @@
+COPY --from=criu             /out/ /usr/local/bin/
+COPY --from=registry         /out/ /usr/local/bin/
+COPY --from=dockercli        /out/ /usr/local/cli/
+COPY --from=frozen-images    /out/ /docker-frozen-images
```
now using the `COPY --link` feature so should not matter anymore.

crazy-max commented:
As discussed bring back `Docker.simple`. Also remove unneeded `XX_PREFER_CROSS` and remove `CC=$(xx-info)-gcc` which is fixed in https://github.com/tonistiigi/xx/releases/tag/v1.1.1.

crazy-max commented:
A new flaky one? https://github.com/moby/moby/runs/6686840944?check_suite_focus=true#step:15:4957

```
=== FAIL: github.com/docker/docker/integration-cli TestDockerSuite/TestRunTwoConcurrentContainers (4.79s)
    docker_cli_run_test.go:809: assertion failed: error is not nil: 
        Command:  D:\a\moby\moby\out\docker.exe run busybox sleep 2
        ExitCode: 125
        Error:    exit status 125
        Stdout:   
        Stderr:   D:\a\moby\moby\out\docker.exe: Error response from daemon: failed to create shim task: hcs::CreateComputeSystem da7623389864a44154755816bdf8e68150abf4553b2fe78b10f01c902f2b2b83: The parameter is incorrect.: unknown.
        
        
        Failures:
        ExitCode was 125 expected 0
        Expected no error
    --- FAIL: TestDockerSuite/TestRunTwoConcurrentContainers (4.79s)
```

Looks to be already identified in https://github.com/moby/moby/pull/42270#issuecomment-820567886

thaJeztah commented on hack/make.sh:
```diff
@@ -83,55 +83,50 @@ add_buildtag() {
-	# Tell cgo the compiler is Clang, not GCC
-	# https://code.google.com/p/go/source/browse/src/cmd/cgo/gcc.go?spec=svne77e74371f2340ee08622ce602e9f7b15f29d8d3&r=e6794866ebeba2bf8818b9261b54e2eef1c9e588#752
-	export CC=clang
+if ${PKG_CONFIG} 'libsystemd >= 209' 2> /dev/null; then
+	DOCKER_BUILDTAGS+=" journald"
+elif ${PKG_CONFIG} 'libsystemd-journal' 2> /dev/null; then
+	DOCKER_BUILDTAGS+=" journald journald_compat"
+fi
```
I'm actually wondering if we still need this at all. Looks like this was added in https://github.com/moby/moby/commit/6cdc4ba6cd5178037466c50ebe03a7eb111c43b1 (https://github.com/moby/moby/pull/20617) in 2016, and likely because at the time we were still building for CentOS 6.

doing some quick checks;

```bash
docker run -it --rm centos:7

yum install -y systemd-devel

pkg-config 'libsystemd >= 209' && echo "OK" || echo "KO"
OK

pkg-config --print-provides 'libsystemd'
libsystemd = 219

pkg-config --print-provides 'libsystemd-journal'
libsystemd-journal = 219
```

And on a `debian:buster` (oldstable)

```bash
docker run -it --rm debian:buster

apt-get update && apt-get install -y libsystemd-dev pkg-config

pkg-config 'libsystemd >= 209' && echo "OK" || echo "KO"
OK

pkg-config --print-provides 'libsystemd'
libsystemd = 241

# only for old distros (now part of libsystemd-dev); https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=795147
pkg-config --print-provides 'libsystemd-journal'
Package libsystemd-journal was not found in the pkg-config search path.
Perhaps you should add the directory containing `libsystemd-journal.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libsystemd-journal' found
```

OpenSUSE leap (I think that's built for s390x)

```bash
docker run -it --rm docker.io/opensuse/leap:15

zypper install -y systemd-devel

pkg-config 'libsystemd >= 209' && echo "OK" || echo "KO"
OK

pkg-config --print-provides 'libsystemd'
libsystemd = 246

pkg-config --print-provides 'libsystemd-journal'
Package libsystemd-journal was not found in the pkg-config search path.
Perhaps you should add the directory containing `libsystemd-journal.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libsystemd-journal' found
```

thaJeztah commented on testutil/fixtures/load/frozen.go:
```diff
@@ -82,28 +82,36 @@ func imageExists(client client.APIClient, name string) bool {
+			return err
+		}
+		err = func(tarfile fs.FileInfo) error {
+			reader, err := os.OpenFile(path.Join(frozenImgDir, tarfile.Name()), os.O_RDONLY, 0644)
```
Should probably use `filepath.Join`, not `path.Join` (I also thought "path" is for Linux, and "filepath" for "portable", but turns out `path` is nowadays advertised to only be used for URL(like) things.

thaJeztah reviewed (commented):
Only went through the first commit (sorry, lots of questions; let me know if they make sense, and I don't have clearcut answers to all of them 😅)

crazy-max commented on .dockerignore:
```diff
@@ -1,4 +1,3 @@
-.git
```
> This would also mean we'd invalidate the cache on any change in the .git directory (which for me happens a lot, e.g. add / remove branches)

Yes indeed cache would be invalidated quite often to load the context here: https://github.com/moby/moby/pull/43529/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R656. I have checked on ci (fetch depth 0) and loading the context with `.git` + building dockerd and docker-proxy takes ~30sec: https://github.com/moby/moby/runs/7425238835?check_suite_focus=true#step:6:2230. Locally it's even faster for me.

> Besides potential metadata (which commit did we build from for inclusion in docker version), are there other reasons we would need this?

Yes this is just to invoke git in https://github.com/moby/moby/pull/43529/files#diff-9eace97aaee3a26c826f37605437498f930565a99a4f072cc476bfeac6199e73R28-R36 if `VERSION`, `GITCOMMIT` or `BUILDTIME` is empty.

> ^^ as, in general, the build should not depend on git (i.e., it should be possible to build from a source tarball, which could be downloaded from the releases page)

Yes we could only rely on ARGs to set `VERSION`, `GITCOMMIT` and `BUILDTIME` for sandboxed builds.

thaJeztah replied:
> Yes this is just to invoke git in https://github.com/moby/moby/pull/43529/files#diff-9eace97aaee3a26c826f37605437498f930565a99a4f072cc476bfeac6199e73R28-R36 if VERSION, GITCOMMIT or BUILDTIME is empty.

(let me copy that code, as the link will break on rebase)

```bash
if [ -z "$VERSION" ]; then
 	VERSION=$(git describe --match 'v[0-9]*' --dirty='.m' --always --tags | sed 's/^v//' 2> /dev/null || echo "unknown-version")
 fi
 if [ -z "$GITCOMMIT" ]; then
 	GITCOMMIT=$(git rev-parse --short HEAD 2> /dev/null || true)
 fi
 if [ -z "$BUILDTIME" ]; then
 	BUILDTIME=$(date -u --date="@${SOURCE_DATE_EPOCH:-$(date +%s)}" +"%Y-%m-%dT%H:%M:%SZ")
 fi
```

> Yes we could only rely on ARGs to set VERSION, GITCOMMIT and BUILDTIME for sandboxed builds.

Yeah, leaning toward that, as;

- we could already (and already do?) pass that information as argument
- 30 seconds is still 30 seconds 😅 
- for local builds, I don't think the information is really relevant (`dev` as "version" is fine, and I know I just built a binary, so I don't really care about build-time)


Of course happy to hear other opinions

cpuguy83 replied:
`COPY --ignore` would be real nice :)

I think for local dev we'd typically have the source mounted in anyway, in which case we skip the `COPY` step anyway.

crazy-max commented on Dockerfile:
```diff
@@ -1,294 +1,550 @@
-ARG SYSTEMD="false"
+ARG BASE_IMAGE="debian:bullseye"
 ARG GO_VERSION=1.18.3
+ARG XX_VERSION=1.1.1
```
Yes good point, will move it under dev deps

crazy-max commented on Dockerfile:
```diff
@@ -1,294 +1,550 @@
+# dummy stage to make sure the image is built for unsupported deps
+FROM --platform=$BUILDPLATFORM busybox AS build-dummy
+RUN mkdir -p /out
+FROM scratch AS binary-dummy
+COPY --from=build-dummy / /
```
Oh the `COPY` command is wrong will fix that. fwiw the `binary-dummy` scratch image only needs the `/out` folder so incompatible platforms to build a tool can still copy and not trigger an error.

crazy-max commented on Dockerfile:
```diff
@@ -1,294 +1,550 @@
+      lld \
+      make \
+      pkg-config
+COPY --from=golang /usr/local/go /usr/local/go
+ENV GOROOT="/usr/local/go"
+ENV GOPATH="/go"
+ENV PATH="$GOPATH/bin:/usr/local/go/bin:$PATH"
+RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"
```
This is a follow-up of https://github.com/moby/moby/pull/43553#issuecomment-1116067966. Basically we would need another base image to support the `riscv64` architecture: `ubuntu:22.04`.

As `golang` does not support an ubuntu flavor we need to make sure the ubuntu and debian base images coexist in the Dockerfile otherwise we can't build anymore against `armel`.

This has been tested in https://github.com/crazy-max/moby/blob/a4990cbf40eb4be78d704feb368e9d978756bf6e/Dockerfile#L3-L6 so we can reuse the same stage without changes in https://github.com/crazy-max/moby/blob/a4990cbf40eb4be78d704feb368e9d978756bf6e/Dockerfile#L73-L97.

This is just to avoid drastic changes in a follow-up. Sorry I should have clarify why we would need this.

thaJeztah replied:
> This is a follow-up of https://github.com/moby/moby/pull/43553#issuecomment-1116067966. Basically we would need another base image to support the riscv64 architecture: ubuntu:22.04.

Ah, hm.. do we know what's missing in Debian? Are there backport / testing debian repositories with packages we need, or is this "newer version of (g)cc etc?"

Trying to see what we could do to simplify things (especially because they're for a still somewhat "esoteric" platform)

crazy-max commented on Dockerfile:
```diff
@@ -1,294 +1,550 @@
+ENV GO111MODULE=off
 
-FROM --platform=linux/amd64 base AS cross-true
+FROM containerd-base-${DOCKER_LINKMODE} AS containerd-build
```
Sure we can do that too but I think it's more elegant and readable that way.

thaJeztah replied:
Hm, right, but now (I _think_) we're building / installing the shared bits _twice_, and I have to "scroll up" to learn what environment variables are set (which are used as part of the build / compile).

I think there's value in keeping those environment variables local / close to where they're used, so that I can see at a glance what's used 🤔

crazy-max replied:
Fair enough I will move them in `RUN` if you prefer :+1:

crazy-max commented on Dockerfile:
```diff
@@ -1,294 +1,550 @@
+ARG DELVE_VERSION=v1.8.1
+ARG CRIU_VERSION=v3.16.1
+ARG CRUN_VERSION=1.4.5
+ARG REGISTRY_VERSION=v2.3.0
+ARG REGISTRY_VERSION_SCHEMA1=v2.1.0
```
I can bring them back sure

crazy-max commented on Dockerfile:
```diff
@@ -1,294 +1,550 @@
+RUN --mount=from=dockercli-src,src=/usr/src/dockercli/components/cli,rw \
+    --mount=type=cache,target=/root/.cache \
+    --mount=type=cache,target=/go/pkg/mod <<EOT
+  set -e
+  go build -o /out/docker -v ./cmd/docker
+  xx-verify /out/docker
```
> also be building with the currently used Go version

Good point I will make some enhancements

crazy-max replied:
Done

crazy-max commented on Dockerfile:
```diff
@@ -1,303 +1,550 @@
-# attaching debugger to it.
-#
-ARG DELVE_VERSION=v1.8.1
-# Delve on Linux is currently only supported on amd64 and arm64;
-# https://github.com/go-delve/delve/blob/v1.8.1/pkg/proc/native/support_sentinel.go#L1-L6
```
Yes this is still the case, I will bring back that comment and make this stage conditional.

thaJeztah commented on Dockerfile:
```diff
@@ -1,294 +1,550 @@
+# containerd
+FROM base AS containerd-src
+WORKDIR /usr/src/containerd
+RUN git clone https://github.com/containerd/containerd.git .
+ARG CONTAINERD_VERSION
+RUN git fetch origin && git checkout -q "$CONTAINERD_VERSION"
```
Yeah, perhaps. Although if we copy it to one of the base images, it would only be invalidated if that script itself changed, correct?

I was curious; did a quick try; looks to be faster (but I guess, it depends also on how many of these we're running) (would be nice if the output showed size as well, besides timing)

```dockerfile
# syntax=docker/dockerfile:1

ARG CONTAINERD_VERSION=v1.6.6

FROM golang:1.18-bullseye AS golang

FROM golang AS clone
WORKDIR /usr/src/containerd
ARG CONTAINERD_VERSION
RUN git clone https://github.com/containerd/containerd.git . \
 && git fetch origin && git checkout -q "$CONTAINERD_VERSION"
 
FROM golang AS shallow1
WORKDIR /usr/src/containerd
ARG CONTAINERD_VERSION
RUN <<EOF
set -ex

git init .
git remote add origin "https://github.com/containerd/containerd.git"

REF="$CONTAINERD_VERSION"
REF_FETCH="$REF"
# if ref is branch or tag, retrieve its canonical form
REF=$(git ls-remote --refs --heads --tags origin "$REF" | awk '{print $2}')
if [ -n "$REF" ]; then
    # if branch or tag then create it locally too
    REF_FETCH="$REF:$REF"
else
    REF="FETCH_HEAD"
fi
git fetch --update-head-ok --depth 1 origin "$REF_FETCH"
git checkout -q "$REF"
EOF


FROM golang AS shallow2
WORKDIR /usr/src/containerd
ARG CONTAINERD_VERSION
RUN <<EOF
git init .
git remote add origin "https://github.com/containerd/containerd.git"
git fetch --depth 1 origin "$(CONTAINERD_VERSION)"
git checkout -q FETCH_HEAD
EOF

FROM scratch AS final
COPY --from=clone    /usr/src/containerd/Makefile .
COPY --from=shallow1 /usr/src/containerd/Makefile .
COPY --from=shallow2 /usr/src/containerd/Makefile .
```

```bash
docker build -t foo --no-cache .
[+] Building 14.1s (16/16) FINISHED
 => [internal] load build definition from Dockerfile                                                                                    0.0s
 => => transferring dockerfile: 32B                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                       0.0s
 => => transferring context: 2B                                                                                                         0.0s
 => resolve image config for docker.io/docker/dockerfile:1                                                                              0.9s
 => CACHED docker-image://docker.io/docker/dockerfile:1@sha256:443aab4ca21183e069e7d8b2dc68006594f40bddf1b15bbd83f5137bd93e80e2         0.0s
 => [internal] load build definition from Dockerfile                                                                                    0.0s
 => [internal] load .dockerignore                                                                                                       0.0s
 => [internal] load metadata for docker.io/library/golang:1.18-bullseye                                                                 0.9s
 => [golang 1/1] FROM docker.io/library/golang:1.18-bullseye@sha256:6bc0fe859a2a28af025f8885fc073996dd14d0371d5c0474aed798317849bc18    0.0s
 => => resolve docker.io/library/golang:1.18-bullseye@sha256:6bc0fe859a2a28af025f8885fc073996dd14d0371d5c0474aed798317849bc18           0.0s
 => CACHED [shallow2 1/2] WORKDIR /usr/src/containerd                                                                                   0.0s
 => [clone 2/2] RUN git clone https://github.com/containerd/containerd.git .  && git fetch origin && git checkout -q "v1.6.6"          11.8s
 => [shallow2 2/2] RUN <<EOF (git init ....)                                                                                            6.0s
 => [shallow1 2/2] RUN <<EOF (set -ex...)                                                                                               7.7s
 => [final 1/3] COPY --from=clone    /usr/src/containerd/Makefile .                                                                     0.0s
 => [final 2/3] COPY --from=shallow1 /usr/src/containerd/Makefile .                                                                     0.0s
 => [final 3/3] COPY --from=shallow2 /usr/src/containerd/Makefile .                                                                     0.0s
 => exporting to image                                                                                                                  0.0s
 => => exporting layers                                                                                                                 0.0s
 => => writing image sha256:a22d3ddcbdbc94d5f5f1f484ee40527474be7d4ea46d8a87146955ea7dcf70ff                                            0.0s
 => => naming to docker.io/library/foo                                                                                                  0.0s

```

crazy-max replied:
Yes indeed with `--depth=1` it's faster

thaJeztah commented on .github/workflows/ci.yml:
```diff
@@ -15,17 +15,22 @@ on:
 
 env:
-  BUNDLES_OUTPUT: ./bundles
+  PLATFORM: "Docker Engine - Community"
+  PRODUCT: "Moby"
+  DEFAULT_PRODUCT_LICENSE: "Community Engine"
+  PACKAGER_NAME: "Docker, Inc."
```
Yeah, or even "empty" would probably work (haven't fully checked where they're all used); what did we use before this PR?

crazy-max replied:
Was empty previously but I like to have something so I can double check the built artifacts in CI have the correct resource info (for windows in particular)

thaJeztah commented on Dockerfile:
```diff
@@ -1,303 +1,550 @@
+ENV ARCH=$TARGETARCH
+ENV VARIANT=$TARGETVARIANT
+RUN <<EOT
+  set -e
+  skopeo copy docker://busybox@sha256:95cf004f559831017cdf4628aaf1bb30133677be8702a8c5f2994629f637a209 --additional-tag busybox:latest docker-archive:///out/busybox-latest.tar
+  skopeo copy docker://busybox@sha256:1f81263701cddf6402afe9f33fca0266d9fff379e59b1748f33d3072da71ee85 --additional-tag busybox:glibc docker-archive:///out/busybox-glibc.tar
+  skopeo copy docker://debian@sha256:dacf278785a4daa9de07596ec739dbc07131e189942772210709c5c0777e8437 --additional-tag debian:bullseye-slim docker-archive:///out/debian-bullseye-slim.tar
+  skopeo copy docker://hello-world@sha256:d58e752213a51785838f9eed2b7a498ffa1cb3aa7f946dda11af39286c3db9a9 --additional-tag hello-world:latest docker-archive:///out/hello-world-latest.tar
+  skopeo copy docker://arm32v7/hello-world@sha256:50b8560ad574c779908da71f7ce370c0a2471c098d44d1c8f6b513c5a55eeeb1 --additional-tag arm32v7/hello-world:latest docker-archive:///out/arm32v7-hello-world-latest.tar
```
Looks like building from source could work (not sure if ideal); 

```bash
CGO_ENABLED=0 go install -tags "exclude_graphdriver_devicemapper exclude_graphdriver_btrfs containers_image_openpgp" github.com/containers/skopeo/cmd/skopeo@v1.9.0

ldd /go/bin/skopeo
	not a dynamic executable

skopeo --help
Various operations with container images and container image registries
...
```

crazy-max replied:
Looks good :+1:

crazy-max commented on docs/contributing/ctn-build.md:
```diff
@@ -0,0 +1,33 @@
+docker buildx bake binary-cross
+
+# build binaries for a specific platform
+docker buildx bake --set *.platform=linux/arm64
```
Overrides can only be done using the `--set` flag with bake. Agree that this example is not on point. As I don't define any target here, it will build the default one (`binary`) so we could also use `--set binary.platform=linux/arm64`. You mostly define the target in the override if you're using a group of targets. In our case we don't have any group with multi targets.

crazy-max commented on docs/contributing/ctn-build.md:
```diff
@@ -0,0 +1,33 @@
+
+# build all for the current host platform
+# output to ./bin
+ALL_OUTPUT=./bin docker buildx bake all
```
See https://github.com/moby/moby/pull/43529/commits/5d5e874196e7ea654032e9c6fd8b63017b4696ad

crazy-max commented on docs/contributing/ctn-build.md:
````diff
@@ -0,0 +1,33 @@
+A [bake definition](https://github.com/docker/buildx/blob/master/docs/reference/buildx_bake.md) named `docker-bake.hcl` is in place to ease the build process:
+
+```shell
+# build binaries for the current host platform
+# output to ./bundles/binary by default
+docker buildx bake
````
> Previously, a build would default to Linux (matching the host's architecture, where "host" being "where the daemon runs").

This has not changed, currently `make [binary|dynbinary|cross]` use Docker to build: https://github.com/moby/moby/blob/7624f8aeb1836cd75a0912d9614930e8a59eaf4e/Makefile#L171-L179

We are just using bake as high level build command to simplify the workflow.

> While I can see value in having `buildx bake` as a replacement for `make` to build binaries for the _current_ platform (but in a sandboxed environment), for this project in particular, that's a bit of a weird assumption, because effectively we only support Linux and Windows (and only on a limited set of architectures)

Windows users can't use `make` (except on WSL) but buildx binary is available on Windows so it's a benefit to rely on it I think.

> This default changes the default to (in many cases) cross-compile, and, e.g. on a macOS machine, produces no useful output at all

I'm not sure I understand, you mean `docker buildx bake` will not produce anything on MacOS? Or you mean `docker buildx bake --set *.platform=local` should output something?

Or you mean that building on M1 machine will cross-comp? Because it will not as it will pull the linux/arm64 image and build natively on it.

Can you explain a bit more?

> if I would run it on a "native" platform, then the default would be a static binary (mostly useful for testing) as this repository does not have recipes to build for specific distros (which is all part of the packaging repositories).

This is the current behavior when you type `make` currently:

```
$ make
docker build  --build-arg=GO_VERSION  -f "Dockerfile" --output=bundles/ --target=binary --build-arg VERSION --build-arg DOCKER_GITCOMMIT --build-arg PRODUCT --build-arg PLATFORM --build-arg DEFAULT_PRODUCT_LICENSE --build-arg PACKAGER_NAME .
```

> Overall, I think the _main purpose_ of the binaries, is;
> 
>     * to be used as a dependency for `make dev` / `make shell`; to have a binary available within the "dev" container (`make BIND_DIR=. binary shell`, or within the container, `hack/make.sh binary && make install`)
> 
>     * for CI (running docker-in-docker)
> 
>     * for Windows (doesn't support docker-in-docker); to build the binary and run it as part of CI

Yes the main purpose is to have a better cross compilation support because today the Dockerfile does not support `--platform` with BuildKit to build efficiently against any arch and so we need to rely on obscure env vars `DOCKER_CROSSPLATFORMS`, `CROSS` instead of automatic platform ARGs in the global scope like `TARGETPLATFORM` and so current `cross` stage will build against all platforms which increase build time.

Same for cross pkgs hardcoded so you have to modify the Dockerfile to add support for extra archs: https://github.com/moby/moby/blob/7624f8aeb1836cd75a0912d9614930e8a59eaf4e/Dockerfile#L108-L112 https://github.com/moby/moby/blob/7624f8aeb1836cd75a0912d9614930e8a59eaf4e/Dockerfile#L137-L147

With these changes it's mostly transparent and will help on docker-ce-packaging: https://github.com/docker/docker-ce-packaging/pull/718/commits/dacaeb69fd404bcb3c3215c31bd2fff78b26f2c7 as well as https://github.com/rumpl/moby/pull/21#issuecomment-1189818960

> Currently, this includes "build containerd and runc", which basically shouldn't be part of that (containerd, runc are separate projects, and should be packaged separately like we do for the .deb and .rpm)

Yes the `cross` job in `ci` workflow currently build binaries and extra tools for testing purpose and to make sure nothing is broken. But agree that in a follow-up we should remove containerd, runc, ... and just grab them from their respective location. Same applies on docker-ce-packaging.

> For _non-static_ binaries (`.deb`, `.rpm`) things are somewhat more complicated;
> 
>     * traditional packaging tools for deb and rpm expect a source tarball as input (no dependency on git repo), and building to be done as part of that tooling
> 
>     * they should be built in an environment that matches the target distro (due to dynamic linking, and dependencies (or their names) to differ between distros)

Nothing change you can build dynamic binaries as before in your spec files like in docker-ce-packaging: https://github.com/docker/docker-ce-packaging/pull/718/commits/150d57b7e7d1e61ddac91bea60f5b3fddbff3c78

> 
>     * our pipeline builds those _within_ a docker container (but does not use docker as part of the "build" by the deb/rpm tools)

Yes as it is today. Do you mean you think about integrating packaging of `.deb` and `.rpm` on moby repo directly?

crazy-max replied:
> as well as [rumpl#21 (comment)](https://github.com/rumpl/moby/pull/21#issuecomment-1189818960)

@thaJeztah I updated https://github.com/rumpl/moby/pull/21 to use changes of this PR in https://github.com/rumpl/moby/pull/24 if you're interested to take a look.

thaJeztah commented on Dockerfile.simple:
```diff
@@ -42,14 +42,14 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
+# install dependencies as linked context from main Dockerfile to avoid
+# deduplication. see docker-bake.hcl for more info
+COPY --link --from=tini        /out/ /usr/local/bin/
+COPY --link --from=runc        /out/ /usr/local/bin/
```
Quick comment; we should probably keep the `.simple` one to be compatible with classic builder (and other builders)

crazy-max replied:
> classic builder

Not sure I understand why?

thaJeztah commented:
oh! FYI: https://github.com/moby/moby/pull/43917 is released (comes with runc v1.1.3)

also includes https://github.com/containerd/containerd/pull/7045, to make it easier to do static compiles (when using the containerd makefile)

crazy-max commented:
When it's not Jenkins it's GH :sweat_smile:: https://github.com/moby/moby/runs/7682078905?check_suite_focus=true#step:5:5327

```
#123 exporting cache
#123 preparing build cache for export
#123 preparing build cache for export 85.8s done
#123 ERROR: GitHub Actions is temporarily unavailable. Please visit https://www.githubstatus.com/ for the status of our services.
```

crazy-max commented:
All green @thaJeztah

crazy-max committed:
Dockerfile: add cross compilation helper

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: pin criu version and build from source

pin criu for better reproducibility and build from
source so we can use it across any platform.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: add dummy stage

dummy stage allows to bypass build for deps that
don't support some platforms

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: handle multi base image for riscv64 and armel support

To add support for riscv64 builds we need crossbuild packages for
riscv64 but current golang image with debian bullseye does not
support it. Ubuntu 22.04 supports riscv64 but unfortunately drops
support for armel arch. Therefore we need a multi base image that
will be picked up based on the target platform we want to build.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: runc cross

runc build in Dockerfile is limited to host
platform and could not be cross-built for
other platforms.

this change allows to build against any platforms
if we want to smoke test in a follow-up but also
enhance e2e tests for linux and windows in our
pipeline.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: tini cross

tini build in Dockerfile is limited to host
platform and could not be cross-built for
other platforms.

this change allows to build against any platforms
if we want to smoke test in a follow-up but also
enhance e2e tests for linux and windows in our
pipeline.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: rootlesskit cross

rootlesskit build in Dockerfile is limited to host
platform and could not be cross-built for
other platforms.

this change allows to build against any platforms
if we want to smoke test in a follow-up but also
enhance e2e tests for linux and windows in our
pipeline.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: containerutility cross

containerutility build in Dockerfile is limited to
windows platform atm but enabling cross build for
it enhance and reduce footprint in our piepline
for linux and windows e2e tests.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: verify and better cache for go-swagger stage

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: align deps format and output

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: use TARGETPLATFORM for delve stage and verify

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: verify and better cache for dockercli stage

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: use global scope platform args for registry stage and verify

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: better cache for crun stage

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: GO111MODULE=on by default

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: enhanced for cross compilation

Better support for cross compilation so we can rely
on --platform flag of buildx for a seamless integration.

This removes not necessary extra cross logic in the Dockerfile
as well as hack scripts.

Tried my best to reduce the footprint of changes but modifying
one bit in the Dockerfile involves other changes in ./hack
scripts. Non-sandboxed build invocation is still supported.

It also handles cross compilation for external tools dynamically
based on platform arg available in global scope (containerd,
runc, tini, ...).

Dev stages have been updated accordingly to changes for cross
comp as well as linked tools (swagger, tomll, gotestsum, ...)

The current bake definition has been updated to take the changes
into account as well as gha workflows.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: refactor Dockerfile.simple and create gha workflow

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
riscv64 cross comp support

needs to update supported platforms for pie buildmode
and adds smoke test

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
docs: contributing docs update for cross comp

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: remove hardcoded platform for vpnkit stage

vpnkit stage only supports linux/amd64 and linux/arm64
platforms when building dev image and will crash if we
try building against another platform.

with this change we can still build the dev image
against any platform using dummy scratch base.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max commented:
As discussed I split a bit more so it should be easier to review.

thaJeztah commented:
Haven't looked closely yet at the latest changes but saw CI failing, so just copying output from that.

This looks odd (the `''` after `hack/make.sh`) but perhaps that was always there and I just never noticed;

```bash
docker run -t --privileged -v /home/ubuntu/workspace/moby_PR-43529/bundles/2:/go/src/github.com/docker/docker/bundles -v /home/ubuntu/workspace/moby_PR-43529/bundles/dynbinary-daemon:/go/src/github.com/docker/docker/bundles/dynbinary-daemon -v /home/ubuntu/workspace/moby_PR-43529/.git:/go/src/github.com/docker/docker/.git --name docker-pr68-2 -e KEEPBUNDLE=1 -e TESTDEBUG -e TESTFLAGS -e TEST_SKIP_INTEGRATION -e TEST_SKIP_INTEGRATION_CLI -e DOCKER_GITCOMMIT=18b348d10b26579baeab2506f9b971866b963bcc -e DOCKER_GRAPHDRIVER -e TIMEOUT -e VALIDATE_REPO=https://github.com/moby/moby.git -e VALIDATE_BRANCH=master docker:18b348d10b26579baeab2506f9b971866b963bcc hack/make.sh '' test-integration
```

Actual failure in integration tests (amd64) looks to be;

```bash
2022-08-20T12:52:54.809Z] ---> Making bundle: .integration-daemon-start (in bundles/test-integration)
[2022-08-20T12:52:54.809Z] Using test binary docker
[2022-08-20T12:52:54.809Z] error: binary-daemon or dynbinary-daemon must be run before .integration-daemon-start
[2022-08-20T12:52:55.739Z] + c=1
[2022-08-20T12:52:55.739Z] + for job in $(jobs -p)
[2022-08-20T12:52:55.739Z] + wait 19199
[2022-08-20T12:54:32.136Z] ---> Making bundle: .integration-daemon-start (in bundles/test-integration)
[2022-08-20T12:54:32.136Z] Using test binary docker
[2022-08-20T12:54:32.136Z] error: binary-daemon or dynbinary-daemon must be run before .integration-daemon-start
[2022-08-20T12:54:32.136Z] ---> Making bundle: .integration-daemon-start (in bundles/test-integration)
[2022-08-20T12:54:32.136Z] Using test binary docker
[2022-08-20T12:54:32.136Z] error: binary-daemon or dynbinary-daemon must be run before .integration-daemon-start
```

On `cross`, failure is;

```bash
[2022-08-20T12:47:11.112Z] + docker run --rm -t --privileged -v /home/ubuntu/workspace/moby_PR-43529/bundles:/go/src/github.com/docker/docker/bundles --name docker-pr68 -e DOCKER_GITCOMMIT=0e84d7183b296f8dc56662cc856800d4a7c29e58 -e DOCKER_GRAPHDRIVER docker:0e84d7183b296f8dc56662cc856800d4a7c29e58 hack/make.sh cross
[2022-08-20T12:47:11.381Z] 
[2022-08-20T12:47:11.640Z] Removing bundles/
[2022-08-20T12:47:11.640Z] 
[2022-08-20T12:47:11.640Z] ---> Making bundle: cross (in bundles/cross)
[2022-08-20T12:47:11.640Z] hack/make.sh: line 116: /go/src/github.com/docker/docker/hack/make/cross: No such file or directory
script returned exit code 1
```

crazy-max commented:
@thaJeztah This is expected as Jenkins changes are not propagated because I don't have write access to this repo. Can you rebase https://github.com/moby/moby/pull/43613?

thaJeztah commented:
> @thaJeztah This is expected as Jenkins changes are not propagated because I don't have write access to this repo. Can you rebase #43613?

Oh! Didn't think of that; I'll do a rebase later 👍

thaJeztah commented:
done 👍 🤞

crazy-max commented on Dockerfile:
```diff
@@ -1,21 +1,63 @@
 RUN echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
 ARG APT_MIRROR
 RUN sed -ri "s/(httpredir|deb).debian.org/${APT_MIRROR:-deb.debian.org}/g" /etc/apt/sources.list \
  && sed -ri "s/(security).debian.org/${APT_MIRROR:-security.debian.org}/g" /etc/apt/sources.list
```
I think it's mainly for Jenkins nodes that had issues in the past: https://github.com/moby/moby/blob/464882e39844809b4f0a4ab11392646ec317f407/Jenkinsfile#L25

thaJeztah replied:
Yes, could be indeed; and in the past we were using pet-machines for many, so perhaps we would hit some rate limits. I guess that's lower risk now that Jenkins uses autoscaling nodes.

Not sure if the same could happen on GitHub Actions of course

crazy-max committed:
Dockerfile: use TARGETVARIANT for frozen images script

using TARGETVARIANT in frozen-images stage implies changes in
download-frozen-image-v2.sh script to add support for variants
so ware able to build against more platforms.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: containerd cross

containerd build in Dockerfile is limited to
host platform and could not be cross-built
for other platforms.

this change allows to build against any platforms
if we want to smoke test in a follow-up but also
enhance e2e tests for linux and windows in our
pipeline.

also introduced DOCKER_LINKMODE to be able to
build dynamic or static binaries.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: add "all" stage to build binaries and extra tools

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

thaJeztah commented:
```
hack/make.sh: line 116: /go/src/github.com/docker/docker/hack/make/cross: No such file or directory
script returned exit code 1

crazy-max commented:
> ```
> hack/make.sh: line 116: /go/src/github.com/docker/docker/hack/make/cross: No such file or directory
> script returned exit code 1
> ```

Yes related to no write access to this repo so Jenkinsfile changes not propagated.

crazy-max committed:
Dockerfile: smoke test stage and gha workflow

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max commented:
@neersighted Yes @corhere got the same issue yesterday. Rebasing should work, let's see.

rumpl commented:
We've been using this to bundle dockerd with containerd for months and it works like a charm!

neersighted commented:
I would really like to see this or a version of it merged; as we're basically blocked on 22.06 by this. Granted, this is a monster of a PR, but it's clear some major refactoring of the build has to happen, and no one has proposed a good way to split this work up into more digestible chunks (or rather, @thaJeztah tried with one or two PRs, but it didn't get too far).

corhere commented:
> one has proposed a good way to split this work up into more digestible chunks (or rather, @thaJeztah tried with one or two PRs, but it didn't get too far).

To me, it looks like nobody has really tried to get any split-up work reviewed and merged. #44114 is just sitting there with no "PTAL" pings.

--------

How is 22.06 blocked? For argument's sake, what would the consequences be if we were to abandon this PR and try to ship 22.06 today as-is?

How much of what this PR touches is _required_ to ship 22.06? Can we cherry-pick just that into a PR and leave the rest for follow-ups? I find it hard to believe that we can't ship 22.06 unless we e.g. added a GHA workflow which validates Dockerfile.simple, modified how delve is installed in the dev container image, collected all the dependency version arguments into one block, changed the default in the Dockerfile to `GO111MODULE=on`, or switched `/build` paths to `/out`.

neersighted commented:
As I understand it, 22.06 is blocked on this because cross-compilation as used by docker/ce-packaging is broken and relies on fixes in this PR to fix it. Docker CE is blocking the release of Moby as there is a want to release both (Docker CE binaries and Moby sources) at the same time.

crazy-max commented:
Static packages with current stable (20.10) are indeed not in good shape in docker-ce-packaging but it should not be a blocker for 22.06.

If I recall correctly the idea was first to split static packages as atm docker-ce-cli ships both the cli and some plugins. Same with docker-ce that ships dockerd, runc and containerd. The cross compilation issue for static packages with moby is that `--platform` and `TARGETPLATFORM` is not possible with the current Dockerfile. We also have to download cross packages for all platforms for each build and we are limited to a subset of platforms. This is what this PR tries to solve.

But again not a blocker for me as we are in the same situation as 20.10 but would be great for v-next at least.

sam-thibault commented:
@crazy-max @thaJeztah It sounds like we can move this off the 22.06 milestone?  Or is there part of this change we must include for 22.06?

neersighted commented:
Gonna go ahead and move this to v-next.

crazy-max commented:
I think we can close this one now
